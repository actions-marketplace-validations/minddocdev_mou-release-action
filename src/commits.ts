import { context, GitHub } from '@actions/github';

// Octokit's commit type subset
interface Commit {
  author: {
    login: string,
    html_url: string,
  },
  html_url: string,
  commit: {
    message: string,
  },
  sha: string,
}

const commitGroups: {
  [index: string]: { title: string, commits: Commit[] },
} = {
  feat: {
    title: '**:zap: Features**',
    commits: [],
  },
  fix: {
    title: '**:wrench: Fixes**',
    commits: [],
  },
  docs: {
    title: '**:books: Documentation**',
    commits: [],
  },
  style: {
    title: '**:nail_care: Style changes**',
    commits: [],
  },
  refactor: {
    title: '**:mountain: Refactors**',
    commits: [],
  },
  test: {
    title: '**:traffic_light: Tests**',
    commits: [],
  },
  chore: {
    title: '**:construction: Maintenance**',
    commits: [],
  },
};

const uncategorizedCommits: Commit[] = [];

/**
 * Generate a Markdown string based on commit information.
 * @param commits - The JSON array with Github commit info.
 */
export async function commitParser(github: GitHub, commitDiffBase: string, releaseTag: string) {
  const { owner, repo } = context.repo;
  const compareCommitsResponse = await github.repos.compareCommits({
    owner,
    repo,
    base: commitDiffBase,
    head: releaseTag
  });
  const { data: { commits } } = compareCommitsResponse;

  commits.forEach((commit) => {
    const { message } = commit.commit;

    // Check if commit message matches to any of the defined categories
    let categoryMatch = false;
    Object.keys(commitGroups).some((category) => {
      if (message.startsWith(`${category}:`) || message.startsWith(`${category}(`)) {
        commitGroups[category].commits.push(commit);
        categoryMatch = true;
        return true;
      }
    });
    if (!categoryMatch) uncategorizedCommits.push(commit);
  });

  let commitMd = '';

  const formatCommit = (commit: Commit, stripGroup = false) => {
    const { login: username, html_url: authorUrl } = commit.author;
    let { message } = commit.commit;
    const { sha } = commit;
    if (stripGroup) {
      // Remove group information for changelog
      message = message.split(':')[1].trim();
    }
    // Always capitalize commit messages
    message = `${message[0].toUpperCase()}${message.slice(1)}`;
    commitMd = `${commitMd}- ${message} - [${sha.substring(0, 8)}](${
      commit.html_url
    })([@${username}](${authorUrl}))\n`;
  };

  uncategorizedCommits.forEach((commit) => formatCommit(commit));

  Object.keys(commitGroups).forEach((category) => {
    const { title, commits } = commitGroups[category];
    if (commits.length !== 0) {
      commitMd = `${commitMd}\n${title}\n`;
      commits.forEach(commit => formatCommit(commit, true));
    }
  });

  return commitMd.trim();
}
