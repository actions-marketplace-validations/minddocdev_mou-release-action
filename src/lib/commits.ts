import { context, GitHub } from '@actions/github';
import * as core from '@actions/core';

import { VersionType } from './version';

// Octokit's commit type subset
interface Commit {
  username: string,
  userUrl: string
  commitUrl: string,
  message: string,
  sha: string,
}

/**
 * Generate a Markdown string based on commit information.
 * @param commits - The JSON array with Github commit info.
 */
export async function commitParser(
  github: GitHub,
  baseRef: string,
  taskPrefix: string,
  taskBaseUrl?: string,
  commitScope?: string,
) {
  const commitGroups: {
    [index: string]: { title: string; commits: Commit[] };
  } = {
    // A new feature
    feat: {
      title: '**:zap: Features**',
      commits: [],
    },
    // A bug fix
    fix: {
      title: '**:wrench: Fixes**',
      commits: [],
    },
    // A code change that improves performance
    perf: {
      title: '**:runner: Performance**',
      commits: [],
    },
    // Documentation only changes
    docs: {
      title: '**:books: Documentation**',
      commits: [],
    },
    // Changes that do not affect the meaning of the code (lint changes)
    style: {
      title: '**:nail_care: Style**',
      commits: [],
    },
    // A code change that neither fixes a bug nor adds a feature
    refactor: {
      title: '**:mountain: Refactors**',
      commits: [],
    },
    // Adding missing tests or correcting existing tests
    test: {
      title: '**:traffic_light: Tests**',
      commits: [],
    },
    // Changes that affect the build system or external development dependencies
    chore: {
      title: '**:construction: Maintenance**',
      commits: [],
    },
    // As an alternative to 'chore', but with very similar meaning
    build: {
      title: '**:construction_worker: Build**',
      commits: [],
    },
    // Changes for CI configuration files and scripts (e.g. Github CI, helm values...)
    ci: {
      title: '**:runner: CI**',
      commits: [],
    },
  };
  const uncategorizedCommits: Commit[] = [];

  const changes: string[] = [];
  const tasks: string[] = [];
  const pullRequests: string[] = [];

  const { owner, repo } = context.repo;
  const compareCommitsResponse = await github.repos.compareCommits({
    owner,
    repo,
    base: baseRef,
    head: context.sha,
  });
  const {
    data: { commits },
  } = compareCommitsResponse;

  const categorizeCommit = (commit: Commit) => {
    const { message } = commit;
    // Skip if scope check is required and commit does not have it
    if (commitScope && !message.includes(`(${commitScope}):`)) {
      return;
    }
    // Check if commit message matches to any of the defined categories
    let categoryMatch = false;
    Object.keys(commitGroups).some(category => {
      // Match with or without scope
      if (message.startsWith(`${category}:`) || message.startsWith(`${category}(`)) {
        commitGroups[category].commits.push(commit);
        categoryMatch = true;
        return true;
      }
    });
    if (!categoryMatch) uncategorizedCommits.push(commit);
  };

  const prRegExp = new RegExp('(\\(#\\d+\\))', 'gi');
  const taskRegExp = new RegExp(`(${taskPrefix}\\d+)`, 'gi');
  commits.forEach(githubCommit => {
    const {
      author: { login: username, html_url: userUrl },
      html_url: commitUrl,
      commit: { message },
      sha,
    } = githubCommit;
    const commit: Commit = { username, userUrl, commitUrl, message, sha };

    // Detect if commit is a Github squash. In that case, ignore commit title and convert
    // in multiple single line commits
    if (/\* .*\n/.test(message)) {
      const messageLines = message.split('* ');
      // Retrieve PR link and JIRA information from first line
      const prMatch = prRegExp.exec(messageLines[0]);
      if (prMatch)
        prMatch.slice(1).forEach(pr => pullRequests.push(pr.replace(/(\(|\)|#)/g, '')));
      const taskMatch = taskRegExp.exec(messageLines[0]);
      if (taskMatch) taskMatch.slice(1).forEach(task => tasks.push(task));
      // Categorize all commits except first one
      messageLines
        .slice(1)
        .forEach(messageLine =>
          categorizeCommit({
            username,
            userUrl,
            commitUrl,
            sha,
            message: messageLine.trim(),
          }),
        );
    } else {
      categorizeCommit(commit);
    }
  });

  let changesMd = '';

  const formatCommit = (commit: Commit) => {
    const { username, userUrl, sha, commitUrl } = commit;
    let { message } = commit;
    // Only take into account the commit title
    message = message.split('\n')[0];
    // Detect if commit message has Angular format
    if (/(\w+\(\w+\)|\w+|\(\w+\)):/.test(message)) {
      // Remove group information for changelog (e.g. messages with categories)
      message = message.split(':')[1].trim();
    }
    // Always capitalize commit messages
    message = `${message[0].toUpperCase()}${message.slice(1)}`;
    // Add to global change markdown
    changesMd = `${changesMd}- ${message} - [${sha.substring(
      0,
      8,
    )}](${commitUrl})([@${username}](${userUrl}))\n`;
    // Add to global commit sha list
    changes.push(sha);
  };

  uncategorizedCommits.forEach(formatCommit);

  Object.keys(commitGroups).forEach(category => {
    const { title, commits } = commitGroups[category];
    if (commits.length !== 0) {
      changesMd = `${changesMd}\n${title}\n`;
      commits.forEach(formatCommit);
    }
  });

  core.setOutput('changes', JSON.stringify(changes));
  core.setOutput('tasks', JSON.stringify(tasks));
  core.setOutput('pull_requests', JSON.stringify(pullRequests));

  return {
    changes: changesMd.trim(),
    nextVersionType: VersionType.patch, // TODO - Detect major and minor from commits or PRs
    tasks: tasks
      .map(
        task =>
          `- [${task}](${
            taskBaseUrl ? taskBaseUrl : `https://${owner}.atlassian.net/browse`
          }/${task})\n`,
      )
      .join(),
    pullRequests: pullRequests
      .map(pr => `- [#${pr}](https://github.com/${owner}/${repo}/pull/${pr})\n`)
      .join(),
  };
}
