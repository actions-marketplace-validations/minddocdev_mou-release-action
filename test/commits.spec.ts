/* eslint-disable @typescript-eslint/camelcase */
import { commitParser } from '@minddocdev/mou-release-action/commits';

jest.mock('@actions/github', () => ({
  context: {
    repo: {
      owner: 'theowner',
      repo: 'therepo',
    },
  },
}));

const author = {
  login: 'darioblanco',
  html_url: 'https://authorurl',
};

const html_url = 'https://commiturl';
const sha = '62ec8ea713fdf14e4abaef3d7d5138194dec49ce';

describe('commit', () => {
  test('render commit diff for each category', async () => {
    const compareCommitsResponse = {
      data: {
        commits: [
          {
            author,
            html_url,
            sha,
            commit: { message: 'feat:super feature' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'fix: My fix' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'docs(context): document everything' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'style: awesome style' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'refactor: one does not simply refactor' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'test(: Tests are good' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'chore: somebody has to keep things going' },
          },
          {
            author,
            html_url,
            sha,
            commit: { message: 'uncategorized commit' },
          },
        ],
      },
    };
    const compareCommits = jest.fn(() => compareCommitsResponse);
    const github = { repos: { compareCommits } };
    expect(await commitParser(github as any, 'v1.0.0', 'v1.1.1')).toMatchSnapshot();
  });
});
