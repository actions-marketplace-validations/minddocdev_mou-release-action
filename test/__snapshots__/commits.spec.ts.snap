// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit render commit diff for each category 1`] = `
"- Uncategorized commit - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:zap: Features**
- Super feature - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:wrench: Fixes**
- My fix - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:books: Documentation**
- Document everything - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:nail_care: Style changes**
- Awesome style - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:mountain: Refactors**
- One does not simply refactor - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:traffic_light: Tests**
- Tests are good - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))

**:construction: Maintenance**
- Somebody has to keep things going - [62ec8ea7](https://commiturl)([@darioblanco](https://authorurl))"
`;
